### PR TITLE
Windows fixes

### PR DESCRIPTION
**Pull request by MisterDA**
New batch of fixes!
This in particular duplicates the base Windows image to have mingw and msvc images. The final design still may change.
Incorporates #40.

### CODE REVIEW

File: src-opam/dockerfile_windows.ml [hunk 1]
```diff
@@ -75,16 +98,16 @@ module Cygwin = struct
     }
 
   let run_sh ?(cyg=default) fmt = ksprintf (run {|%s\bin\bash.exe --login -c "%s"|} cyg.root) fmt
+  let run_sh_ocaml_env ?(cyg=default) args fmt = ksprintf (run_sh "ocaml-env exec %s -- %s" args) fmt
```
Comment (dra27) [inline src-opam/dockerfile_windows.ml:101]: The syntax for this is weird (I can never remember it...)
```suggestion
  let run_sh_ocaml_env ?cyg:(_ = default) args fmt = ksprintf (run_sh "ocaml-env exec %s -- %s" args) fmt
```
that's how you specify a default value for an optional parameter which you then don't use without triggering warning 27 😜

File: src-opam/dockerfile_opam.ml [hunk 2]
```diff
@@ -244,7 +244,10 @@ let gen_opam2_distro ?(clone_opam_repo=true) ?arch ?labels d =
      yum_opam2 ?labels ?arch ~yum_workaround ~enable_powertools d ()
   | `Zypper -> zypper_opam2 ?labels ?arch d ()
   | `Pacman -> pacman_opam2 ?labels ?arch d ()
-  | `Cygwin -> cygwin_opam2 ?labels ?arch d ()
+  | `Cygwin ->
+     match arch with
+     | None -> failwith "Windows requires an explicit architecture"
```
Comment (dra27) [inline src-opam/dockerfile_opam.ml:249]: Does it need an explicit arch? If so, this should be `invalid_arg`, not `failwith`

File: src-opam/dockerfile_windows.ml [hunk 3]
```diff
@@ -20,6 +20,13 @@ open Printf
 
 let run_cmd fmt = ksprintf (run "cmd /S /C %s") fmt
 let run_powershell fmt = ksprintf (run {|powershell -Command "%s"|}) fmt
+let run_vc ~arch fmt =
+  let arch = match arch with
+    | `I386 -> "x86" | `X86_64 -> "amd64"
+    | `Aarch64 | `Aarch32 | `Ppc64le -> failwith "Cross-compiling isn't supported."
```
Comment (dra27) [inline src-opam/dockerfile_windows.ml:26]: Cross-compiling is the wrong message here - it's an unsupported architecture. I have built OCaml on ARM64 Windows 10 using those tools - it's not cross-compilation!

File: src-opam/dockerfile_windows.ml [hunk 4]
```diff
@@ -47,18 +54,34 @@ let install_visual_studio_build_tools ?(vs_version="16") ?(split=false) componen
...
 
-let append_path paths =
+let prepend_path paths =
   let paths = String.concat ";" paths in
   run {|for /f "tokens=1,2,*" %%a in ('reg query "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /V Path ^| findstr /r "^[^H]"') do `
-       reg add "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /V Path /t REG_EXPAND_SZ /f /d "%%c;%s"|} paths
+        reg add "HKLM\SYSTEM\CurrentControlSet\Control\Session Manager\Environment" /V Path /t REG_EXPAND_SZ /f /d "%s;%%c"|} paths
+
+type windows_port = [
+  | `Cygwin
```
Comment (dra27) [inline src-opam/dockerfile_windows.ml:63]: `` `Cygwin`` is a Unix distribution, it shouldn't be here. I'm anticipating that we'll make it work by providing an equivalent OS base image (say to the Ubuntu ones) which just happens itself to be based on Windows.

File: src-opam/dockerfile_windows.ml [hunk 5]
```diff
@@ -47,18 +54,34 @@ let install_visual_studio_build_tools ?(vs_version="16") ?(split=false) componen
...
-    | `X86_64 -> Some "mingw64c"
-    | _ -> None
-  else None
+    | `X86_64 | `I386 -> begin
+        match port with
+        | `Cygwin -> raise (Invalid_argument "Cygwin port is unsupported")
+        | `Mingw -> Some ("mingw" ^ bitness)
+        | `Msvc -> Some ("msvc" ^ bitness)
+      end
+    | _ -> raise (Invalid_argument "Unsupported architecture")
```
Comment (dra27) [inline src-opam/dockerfile_windows.ml:78]:
```suggestion
    | _ -> invalid_arg "Unsupported architecture"
```

File: src-opam/dockerfile_windows.ml [hunk 6]
```diff
@@ -120,23 +143,24 @@ module Cygwin = struct
   let msvc_packages ?(extra=[]) () = "make" :: "diffutils" :: extra
   let ocaml_for_windows_packages ?cyg ?(extra=[]) ?version:(version="0.0.0.2") () =
     let packages = "make" :: "diffutils" :: "mingw64-x86_64-gcc-g++" :: "vim" :: "git"
-                   :: "curl" :: "rsync" :: "unzip" :: "patch" :: "m4" :: extra in
+                   :: "curl" :: "rsync" :: "unzip" :: "patch" :: "m4" :: "gawk" :: extra in
```
Comment (dra27) [inline src-opam/dockerfile_windows.ml:146]: gawk is a base package, so this shouldn't be necessary?

**Review by dra27 — COMMENTED** [review]
#40's on its way in, so you can rebase without that commit with a few bits and bobs here there. Definitely skip Cygwin in this part for now - the aim is to spin out Cygwin into a container of its own which we won't maintain here (but which we may well then go on to _consume_ for the mingw/msvc images)

File: src-opam/dockerfile_opam.ml (hunk 2 above)
Comment (MisterDA) [reply src-opam/dockerfile_opam.ml:249]: iirc it's for the `ocaml-env` script. Although it claims to be able to detect 64 or 32 bits platforms, I've had cases where it defaulted to 32 bits and wouldn't load the correct compiler. So in this case, I'm requiring an explicit arch to be able to pass `--32` or `--64` later on.
I'll change to `invalid_arg`.

File: src-opam/dockerfile_windows.ml (hunk 1 above)
Comment (MisterDA) [reply src-opam/dockerfile_windows.ml:101]: Weird, I did not see the warning trigger. Noted, I didn't know that syntax. I even learned the shortcut `?(arg=value)` here :)

Comment (MisterDA) [reply src-opam/dockerfile_windows.ml:101]: Oh, but it's a definitely a bug that the parameter is not used. It should be passed to the nested `run_sh` function.

Comment (dra27) [reply src-opam/dockerfile_windows.ml:101]: Ah, so it is - I didn't stare at it hard enough to spot that!

**Pull request merged**